driver/src/orderbook/streamed/updating_orderbook.rs

```diff
@@ -12,10 +16,52 @@ use futures::{
 use orderbook::Orderbook;
 use std::sync::{Arc, Mutex};
 
-// TODO:
-// `pub struct UpdatingOrderbook` that also owns an `Arc<Mutex<Orderbook>>` and creates a
-// background thread running `UpdatingOrderbookThread::update_with_events`.
-// This way `UpdatingOrderbookThread` is independently testable.
+/// An event based orderbook that automatically updates itself with new events from the contract.
+#[derive(Debug)]
+pub struct UpdatingOrderbook {
+    orderbook: Arc<Mutex<Orderbook>>,
+    // When this struct is dropped this sender will be dropped which makes the updater thread stop.
+    _channel: oneshot::Sender<()>,
+}
+
+impl UpdatingOrderbook {
+    pub fn new(
+        _contract: &impl StableXContract,
+        block_timestamp: impl BlockTimestamp + Send + 'static,
+    ) -> Self {
+        let orderbook = Arc::new(Mutex::new(Orderbook::default()));
+        let orderbook_clone = orderbook.clone();
+        let (sender, receiver) = oneshot::channel();
+        // Create stream first to make sure we do not miss any events between it and past events.
+        // TODO: use the real functions once they are implemented
+        let stream = futures::stream::iter(vec![]).boxed(); // contract.stream_events();
+        let past_events = futures::future::ready(Ok(Vec::new())).boxed(); // contract.past_events();
+
+        std::thread::spawn(move || {
```

Comment:
Maybe it makes sense if this is done in two parts - first the past events need to be retrieved, and then the future events, otherwise it doesn't really make sense to get the orderbook while its still catching up on all the past events.

Reply:
Implemented something like this with a bool that gets set when orderbooks is caught up.